File: shorttext/src/main/java/com/tabjin/shorttext/MainActivity.java
package com.tabjin.shorttext;

import android.Manifest;
import android.app.Activity;
import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.ContentResolver;
import android.content.ContentValues;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.provider.ContactsContract;
import android.provider.Telephony;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.telephony.SmsManager;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MainActivity extends AppCompatActivity {

    private Button sendBtn;

    private  PendingIntent deliverPI;
    private  PendingIntent sentPI;
    private  String defaultSmsApp;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if(Build.VERSION.SDK_INT>=23){
            if(checkSelfPermission(Manifest.permission.READ_CONTACTS)==PackageManager.PERMISSION_GRANTED) {
                testGetAllContact();
            }else{
                String[] permission = {
                        Manifest.permission.READ_CONTACTS
                };
                requestPermissions(permission,11);
            }
        }else{
            testGetAllContact();
        }
        defaultSmsApp =  Telephony.Sms.getDefaultSmsPackage(this);
        sendBtn = findViewById(R.id.btn_send);
        sendBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
               /* if(Build.VERSION.SDK_INT>=23){
                    if(checkSelfPermission(Manifest.permission.SEND_SMS)==PackageManager.PERMISSION_GRANTED) {
                        sendSMS("dsafdfdfdsfdsafdsafdsafdsafdsafdsa","15918522240");
                    }else{
                        String[] permission = {
                                Manifest.permission.SEND_SMS
                        };
                        requestPermissions(permission,22);
                    }
                }else{
                    sendSMS("dsafdfdfdsfdsafdsafdsafdsafdsafdsa","15918522240");
                }*/
                setDefault(BuildConfig.APPLICATION_ID);
                TestInsertSMS();
            }
        });

        //处理返回的发送状态
        String SENT_SMS_ACTION = "SENT_SMS_ACTION";
        Intent sentIntent = new Intent(SENT_SMS_ACTION);
        sentPI= PendingIntent.getBroadcast(this, 0, sentIntent,
                0);
        // register the Broadcast Receivers
        this.registerReceiver(new BroadcastReceiver() {
            @Override
            public void onReceive(Context _context, Intent _intent) {
                switch (getResultCode()) {
                    case Activity.RESULT_OK:
                        Toast.makeText(MainActivity.this,
                                "短信发送成功", Toast.LENGTH_SHORT)
                                .show();
                        break;
                    case SmsManager.RESULT_ERROR_GENERIC_FAILURE:
                        break;
                    case SmsManager.RESULT_ERROR_RADIO_OFF:
                        break;
                    case SmsManager.RESULT_ERROR_NULL_PDU:
                        break;
                }
            }
        }, new IntentFilter(SENT_SMS_ACTION));

        //处理返回的接收状态
        String DELIVERED_SMS_ACTION = "DELIVERED_SMS_ACTION";
        // create the deilverIntent parameter
        Intent deliverIntent = new Intent(DELIVERED_SMS_ACTION);
        deliverPI = PendingIntent.getBroadcast(this, 0,
                deliverIntent, 0);
        this.registerReceiver(new BroadcastReceiver() {
            @Override
            public void onReceive(Context _context, Intent _intent) {
                Toast.makeText(MainActivity.this,
                        "收信人已经成功接收", Toast.LENGTH_SHORT)
                        .show();
            }
        }, new IntentFilter(DELIVERED_SMS_ACTION));
    }

    private void setDefault(String name) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            Intent smsIntent = new Intent(Telephony.Sms.Intents.ACTION_CHANGE_DEFAULT);
            smsIntent.putExtra(Telephony.Sms.Intents.EXTRA_PACKAGE_NAME,name);
            startActivity(smsIntent);
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if(grantResults[0]==PackageManager.PERMISSION_GRANTED&&permissions[0].equals(Manifest.permission.READ_CONTACTS)){
            testGetAllContact();
        }
        if(grantResults[0]==PackageManager.PERMISSION_GRANTED&&permissions[0].equals(Manifest.permission.SEND_SMS)){
            sendSMS("dsafdfdfdsfdsafdsafdsafdsafdsafdsa","15918522240");
        }
    }

    public void sendSMS(String content,String phone){
        SmsManager smsManager = SmsManager.getDefault();
        // 拆分短信内容（手机短信长度限制）
        List<String> divideContents = smsManager.divideMessage(content);
        for (String text : divideContents) {
            smsManager.sendTextMessage(phone, null, text, sentPI,
                    deliverPI);
        }


    }

    /**
     * 读取联系人
     */
    public void testGetAllContact()
    {
        //获取联系人信息的Uri
        Uri uri = ContactsContract.Contacts.CONTENT_URI;
        //获取ContentResolver
        ContentResolver contentResolver = this.getContentResolver();
        //查询数据，返回Cursor
        Cursor cursor = contentResolver.query(uri, null, null, null, null);
        List<Map<String,Object>> list = new ArrayList<Map<String,Object>>();
        while(cursor!=null&&cursor.moveToNext())
        {
            Map<String,Object> map = new HashMap<String,Object>();
            StringBuilder sb = new StringBuilder();
            //获取联系人的ID
            String contactId = cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts._ID));
            //获取联系人的姓名
            String name = cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts.DISPLAY_NAME));
            //构造联系人信息
            sb.append("contactId=").append(contactId).append(",Name=").append(name);
            map.put("name", name);
            String id = cursor.getString(cursor.getColumnIndex(ContactsContract.Contacts._ID));//联系人ID


            //查询电话类型的数据操作
            Cursor phones = contentResolver.query(ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                    null,
                    ContactsContract.CommonDataKinds.Phone.CONTACT_ID +" = "+ contactId,
                    null, null);
            while(phones!=null&&phones.moveToNext())
            {
                String phoneNumber = phones.getString(phones.getColumnIndex(
                        ContactsContract.CommonDataKinds.Phone.NUMBER));
                //添加Phone的信息
                sb.append(",Phone=").append(phoneNumber);
                map.put("mobile", phoneNumber);
            }
            phones.close();


            //查询Email类型的数据操作
         /*   Cursor emails = contentResolver.query(ContactsContract.CommonDataKinds.Email.CONTENT_URI,
                    null,
                    ContactsContract.CommonDataKinds.Email.CONTACT_ID + " = " + contactId,
                    null, null);
            while (emails.moveToNext())
            {
                String emailAddress = emails.getString(emails.getColumnIndex(
                        ContactsContract.CommonDataKinds.Email.DATA));
                //添加Email的信息
                sb.append(",Email=").append(emailAddress);
                Log.e("emailAddress", emailAddress);
                map.put("email", emailAddress);


            }
            emails.close();*/
            //Log.i("=========ddddddddddd=====", sb.toString());

            //查询==地址==类型的数据操作.StructuredPostal.TYPE_WORK
            /*Cursor address = contentResolver.query(ContactsContract.CommonDataKinds.StructuredPostal.CONTENT_URI,
                    null,
                    ContactsContract.CommonDataKinds.StructuredPostal.CONTACT_ID + " = " + contactId,
                    null, null);
            while (address.moveToNext())
            {
                String workAddress = address.getString(address.getColumnIndex(
                        ContactsContract.CommonDataKinds.StructuredPostal.DATA));


                //添加Email的信息
                sb.append(",address").append(workAddress);
                map.put("address", workAddress);
            }
            address.close();*/
            //Log.i("=========ddddddddddd=====", sb.toString());

            //查询==公司名字==类型的数据操作.Organization.COMPANY  ContactsContract.Data.CONTENT_URI
           /* String orgWhere = ContactsContract.Data.CONTACT_ID + " = ? AND " + ContactsContract.Data.MIMETYPE + " = ?";
            String[] orgWhereParams = new String[]{id,
                    ContactsContract.CommonDataKinds.Organization.CONTENT_ITEM_TYPE};
            Cursor orgCur = contentResolver.query(ContactsContract.Data.CONTENT_URI,
                    null, orgWhere, orgWhereParams, null);
            if (orgCur.moveToFirst()) {
                //组织名 (公司名字)
                String company = orgCur.getString(orgCur.getColumnIndex(ContactsContract.CommonDataKinds.Organization.DATA));
                //职位
                String title = orgCur.getString(orgCur.getColumnIndex(ContactsContract.CommonDataKinds.Organization.TITLE));
                sb.append(",company").append(company);
                sb.append(",title").append(title);
                map.put("company", company);
                map.put("title", title);
            }
            orgCur.close();*/
            list.add(map);
            Log.i("=========orgName=====", sb.toString());//查看所有的数据
            Log.e("=========map=====", map.toString());//有很多数据的时候，只会添加一条  例如邮箱，
        }

        Log.i("=========list=====", list.toString());//
        cursor.close();
    }

    private void TestInsertSMS() {
        final String ADDRESS = "address";
        final String DATE = "date";
        final String READ = "read";
        final String STATUS = "status";
        final String TYPE = "type";
        final String BODY = "body";
        int MESSAGE_TYPE_INBOX = 1;
        int MESSAGE_TYPE_SENT = 2;
        ContentValues values = new ContentValues();
        /* 手机号 */
        values.put(ADDRESS, "17681823314");
        /* 时间 */
        values.put(DATE, System.currentTimeMillis());
        values.put(READ, 1);
        values.put(STATUS, -1);
        /* 类型1为收件箱，2为发件箱 */
        values.put(TYPE, 2);
        /* 短信体内容 */
        values.put(BODY, "测试插入一条短信");
        /* 插入数据库操作 */
        Uri inserted = getContentResolver().insert(Uri.parse("content://sms"),
                values);
        setDefault(defaultSmsApp);
    }
}

File: dahh/src/main/java/com/tabjin/dahh/MainActivity.kt
package com.tabjin.dahh

import android.Manifest
import android.app.Activity
import android.content.*
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.LinearLayoutManager
import android.text.Editable
import android.text.TextUtils
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.Toast
import com.tabjin.dahh.bean.Contact
import kotlinx.android.synthetic.main.activity_main.*
import android.provider.Telephony
import android.support.v7.app.AlertDialog
import com.tabjin.dahh.model.*


class MainActivity : AppCompatActivity(),View.OnClickListener {

    override fun onClick(v: View?) {
        when(v?.id){
            R.id.tv_send->{
                var num = 0
                for(i in contacts.indices){
                    if(contacts[i].checkstate){
                        num++
                    }
                }
                AlertDialog.Builder(this@MainActivity).setTitle("请确认收件人数")
                            .setMessage("共$num 位收件人，确认发送？")
                            .setCancelable(true)
                            .setPositiveButton("发送") { dialog, which ->
                                dialog.dismiss()
                                tv_stop.visibility = View.VISIBLE
                                tv_send.visibility = View.GONE
                                progressbar.visibility = View.VISIBLE
                                content = et_input.text.toString()
                                startSend()
                            }.setNegativeButton("取消") { dialog, _ -> dialog.dismiss() }
                            .show()
            }
            R.id.tv_stop->{
                tv_stop.visibility = View.GONE
                tv_send.visibility = View.VISIBLE
                progressbar.visibility = View.INVISIBLE
            }
        }
    }

    val SENT_SMS_ACTION = "SENT_SMS_ACTION"
    val DELIVERED_SMS_ACTION = "DELIVERED_SMS_ACTION"
    val TAG = "sendsms"

    var stop = false

    var contacts:ArrayList<Contact> = ArrayList<Contact>()
    var mAdapter:ContactAdapter? = null
    var contactModel:IContact? = null
    var contentCombain: IContent? = null

    var content = ""

    var sendingIndex = -1

    var receiver = object : BroadcastReceiver() {
        override fun onReceive(_context: Context, _intent: Intent) {
            val index = sendingIndex
            if (index <= -1 || index >= contacts.size) {
                return
            }
            Log.e(TAG,"action=${_intent.action}   index = $index")
            when (_intent.action) {
                SENT_SMS_ACTION -> {
                    contacts[index].sentNum++
                    when (resultCode) {
                        Activity.RESULT_OK -> {
                            contacts[index].sentSuccess++
                        }
                        else->{
                            contacts[index].sentFaild++
                        }
                    }

                    if(contacts[index].sentNum == contacts[index].perNum){
                        val nextIndex = nextPosition(contacts,index)
                        //所有拆分短信发送动作完成，进行下一个短信
                        if(nextIndex>-1&&nextIndex<contacts.size){
                            //判断有没有暂停
                            if(progressbar.visibility == View.VISIBLE) {
                                Log.e(TAG,"addTask sendSMS $nextIndex")
                                SMSsendModel.instance.submitTask(this@MainActivity,
                                        contentCombain?.combianContent(content, contacts[nextIndex])!!, nextIndex, contacts)
                                tv_send_name.text = contacts[nextIndex].contactName
                                sendingIndex = nextIndex
                            }
                        }else if(nextIndex == -1){    //所有任务发送完成，
                            progressbar.visibility = View.INVISIBLE
                            tv_send_name.text = "任务完成"
                            tv_stop.visibility= View.GONE
                            tv_send.visibility = View.VISIBLE
                        }
                        //所有拆分后的短信都发送成功，插入短信数据库
                        if(contacts[index].sentSuccess == contacts[index].perNum){
                            Log.e(TAG,"${contacts[index].contactName} 发送完成并成功")
                            ThreadHelper.instance.addInsertTask(this@MainActivity,
                                    contacts[index].contactPhone,
                                    contentCombain?.combianContent(content, contacts[index])!!,
                                    2)
                            contacts[index].sendsuccess = "发送成功"
                            mAdapter?.notifyDataSetChanged()
                        }else if(contacts[index].sentFaild == contacts[index].perNum){
                            contacts[index].sendsuccess = "发送失败"
                            mAdapter?.notifyDataSetChanged()
                        }else{
                            contacts[index].sendsuccess = "部分发送失败"
                            mAdapter?.notifyDataSetChanged()
                        }
                    }
                }
                DELIVERED_SMS_ACTION->{
                    if(index>-1&&index<contacts.size){
                        contacts[index].delevedsuccess = true
                    }
                }
            }

        }
    }

    /**
     * 找到下一个需要发送的联系人
     *
     */
    fun nextPosition(contacts:List<Contact>,index:Int):Int{
        if(index<-1||index>contacts.size-2){
            return -1
        }
        for(i in index+1 until contacts.size){
            if(contacts[i].checkstate){
                return i
            }
        }
        return -1
    }

    var sentNum = 0
    var defaultSmsApp = ""
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        mAdapter = ContactAdapter(this,contacts)
        recyclerview.adapter = mAdapter
        recyclerview.layoutManager = LinearLayoutManager(this,LinearLayoutManager.VERTICAL,false)
        var filter =  IntentFilter(SENT_SMS_ACTION)
//        filter.addAction(DELIVERED_SMS_ACTION)
        var sp = getSharedPreferences(BuildConfig.APPLICATION_ID, Context.MODE_PRIVATE)
        defaultSmsApp = sp.getString("defaultSmsApp","")!!
        if(TextUtils.isEmpty(defaultSmsApp)){
            defaultSmsApp = Telephony.Sms.getDefaultSmsPackage(this)
            //不是本应用，就是系统应用，保存下来
            if(!TextUtils.equals(defaultSmsApp,BuildConfig.APPLICATION_ID)) {
                sp.edit().putString("defaultSmsApp", defaultSmsApp).apply()
            }
        }
        if(!TextUtils.equals(defaultSmsApp,BuildConfig.APPLICATION_ID)) {
            SMSHelper.setDefault(this@MainActivity, BuildConfig.APPLICATION_ID)
        }
        // register the Broadcast Receivers
        this@MainActivity.registerReceiver(receiver, filter)

        contactModel = ContactProviderFactory.instance.getProvider(2)
        contentCombain = ContentProviderFactory.instance.getProvider(1)

        if (Build.VERSION.SDK_INT >= 23) {
            if (checkSelfPermission(Manifest.permission.READ_CONTACTS) != PackageManager.PERMISSION_GRANTED
            ||checkSelfPermission(Manifest.permission.SEND_SMS)!=PackageManager.PERMISSION_GRANTED
            ||checkSelfPermission(Manifest.permission.READ_PHONE_STATE)!=PackageManager.PERMISSION_GRANTED) {
                val permission = arrayOf(Manifest.permission.READ_CONTACTS,Manifest.permission.SEND_SMS,Manifest.permission.READ_PHONE_STATE)
                requestPermissions(permission, 11)
            }
        }

        prefix_input.addTextChangedListener(object :TextWatcher{
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {

            }

            override fun afterTextChanged(s: Editable?) {
                if(progressbar.visibility == View.INVISIBLE&&!TextUtils.isEmpty(s)){
                    getContact(s.toString())
                }else if(TextUtils.isEmpty(s)&&progressbar.visibility == View.INVISIBLE){
                    contacts.clear()
                    mAdapter?.notifyDataSetChanged()
                }
            }
        })


        tv_stop.visibility = View.GONE;
        tv_send.visibility = View.VISIBLE
        tv_stop.setOnClickListener(this)
        tv_send.setOnClickListener(this)

    }

    private fun startSend() {
        sendingIndex = nextPosition(contacts,-1)
        val msg = contentCombain?.combianContent(content,contacts[sendingIndex])!!
        SMSsendModel.instance.continueNum = 0
        SMSsendModel.instance.submitTask(this@MainActivity,msg, sendingIndex, contacts)
        tv_send_name.text = contacts[sendingIndex].contactName

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        for(i in grantResults){
            if(i!=0){
                finish()
                Toast.makeText(this,"请给与相应权限",Toast.LENGTH_SHORT).show()
            }
        }
    }

    fun getContact(prefix:String){
        progressbar.visibility = View.VISIBLE
        ThreadHelper.instance.addTask(Runnable {
            var list: ArrayList<Contact> = contactModel?.getContact(this@MainActivity, prefix)!!
            contacts.clear()
            contacts.addAll(list)
            Log.e("size", "" + list.size)
            runOnUiThread {
                progressbar.visibility = View.INVISIBLE
                mAdapter?.notifyDataSetChanged()
            }
        })
    }

    override fun onDestroy() {
        unregisterReceiver(receiver)

        defaultSmsApp = Telephony.Sms.getDefaultSmsPackage(this)
        if(TextUtils.equals(defaultSmsApp,BuildConfig.APPLICATION_ID)) {
            var sp = getSharedPreferences(BuildConfig.APPLICATION_ID, Context.MODE_PRIVATE)
            var appSms = sp.getString("defaultSmsApp","")
            if(!TextUtils.isEmpty(appSms)) SMSHelper.setDefault(this, defaultSmsApp)
        }
        super.onDestroy()
    }
}